### ISSUE TITLE
Negation and Disjunction from datomic

### ISSUE THREAD

**Issue by narma** [tonsky/datascript#50]
Hello! Thanks for a very good project. Got rid of lots of pain at client-side.

There is [blogpost](http://blog.datomic.com/2015/01/datalog-enhancements.html) about new enchancements in queries, for example disunction (`or`) without using rules and negation, via the new `not` and `not-join` clauses.

I found it very useful and seems it's not too hard to achieve this in datascript.

What do you think about adding this functionality to datascript?

**Comment by tonsky** [comment]
Sure, eventually :)

**Comment by sparkofreason** [comment]
+1

**Comment by bahulneel** [comment]
Is there any progress on this as it would be extremely useful for me?

**Comment by terinjokes** [comment]
Disjoins seem to be implemented, at least as part of 0.15, but I can't get them to work. When I try the following code, adapted from [test/query_or.cljc](https://github.com/tonsky/datascript/blob/e1ca971762401444660924120cc6d9d0be58c7c0/test/datascript/test/query_or.cljc#L12-L30), I get the error "Query for unknown vars: [?e]".

```javascript
var db = ds.db_with(ds.empty_db(), [
  {":db/id": 1, ":name": "Ivan", ":age": 10},
  {":db/id": 2, ":name": "Ivan", ":age": 20},
  {":db/id": 3, ":name": "Oleg", ":age": 10},
  {":db/id": 4, ":name": "Oleg", ":age": 20},
  {":db/id": 5, ":name": "Ivan", ":age": 10},
  {":db/id": 6, ":name": "Ivan", ":age": 20}
]);
ds.q('[:find ?e :where [(or [?e ":name" "Oleg"] [?e ":age" 10])]]', db);
```

Am I doing something wrong, or are "or"s still not ready?

**Comment by bahulneel** [comment]
There's an experimental query namespace called v3 that negation and joins work on, some other things don't though so your mileage may vary.

**Comment by tonsky** [comment]
That’s right, it’s a rewrite of a query engine, it’s not used by default
because it doesn’t do rules, fn/pred calls and probably something else.

**Comment by bahulneel** [comment]
If you just need a disjoint you can can the same effect with two rules with the same head.

**Comment by terinjokes** [comment]
As far as I can tell, query_v3 doesn't get exposed via the JS interop layer. Just to be clear, negation and disjoins only work in it?

@bahulneel Thanks, I saw disjoins were possible beforehand in the blog post linked above. I'm new to datalog (just started 12 hours ago or so), so haven't quite figured how to do write a or-equivalent rule just yet. I'll look into it today.

**Comment by bahulneel** [comment]
Based on your example you could create a rule called `(name-or-age ?name ?age)` with two bodies e.g.:

```clojure
[[(name-or-age ?e ?name ?age)
  [?e :name ?name]]
 [(name-or-age ?e ?name ?age)
  [?e :age ?age]]]
```

Then in the query:
```js
ds.q('[:find ?e :in $ % :where (name-or-age ?e "Oleg" 10)]', db, rules);
```

(not sure what the js native way of making the rules data structure is)

**Comment by terinjokes** [comment]
@bahulneel You put it in quotes like that. :smile:

**Comment by tonsky** [comment]
> As far as I can tell, query_v3 doesn't get exposed via the JS interop
layer. Just to be clear, negation and disjoins only work in it?

Yes. It’s not considered part of DS API in Clojure either. It’s just where
development happens.

**Comment by atroche** [comment]
Hi guys,

I just posted a [$USD300 bounty](https://www.bountysource.com/issues/7853568-negation-and-disjunction-from-datomic) for this issue. Feel free to pitch in with cash or, y'know, claim the bounty =)

**Comment by bahulneel** [comment]
I've shown a method of achieving both using some custom predicates in this ticket: https://github.com/tonsky/datascript/issues/238

**Comment by comnik** [comment]
Sorry for reviving this, but I'd be up for taking a stab at this. @tonsky does it make more sense to wait for / contribute to query_v3 or add negation and disjunction to the current version of the query engine?

**Comment by tonsky** [comment]
I doubt I’ll be able to revive query_v3 any time soon, so more realistic approach would be to add those to the current query impl

**Comment by tonsky** [comment]
Phew. Issue from 2015. Finally implemented not/or/not-join/or-join in 0.17.1. Enjoy!
